### ISSUE TITLE
Twisted IRCClient Remote DoS

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5029]
|<img alt="XiX's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| XiX reported|
|-|-|
|Trac ID|trac#5029|
|Type|defect|
|Created|2011-04-05 21:51:42Z|
|Branch|https://github.com/twisted/twisted/tree/ctcp-dos-5029|

A flaw exists within IRCClient's CTCP implementation.
Multiple CTCP requests can be sent on one line to the client. This can be exploited by a malicious user bundling many erroneous CTCP requests on one line, causing the Twisted client to flood back an errmsg for every request & effectively kill itself due to excess flood.

Example:
```
<XiX> \x01\\x01
<pyn> ERRMSG \\ None: Unknown query '\\'
<XiX> \x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01\x01\\x01
<pyn> ERRMSG \\ None: Unknown query '\\'
<pyn> ERRMSG \\ None: Unknown query '\\'
<pyn> ERRMSG \\ None: Unknown query '\\'
<pyn> ERRMSG \\ None: Unknown query '\\'
<pyn> ERRMSG \\ None: Unknown query '\\'
<pyn> ERRMSG \\ None: Unknown query '\\'
* pyn has quit (Excess Flood)
```

<details><summary>Searchable metadata</summary>

```
trac-id__5029 5029
type__defect defect
reporter__XiX XiX
priority__high high
milestone__ 
branch__branches_ctcp_dos_5029 branches/ctcp-dos-5029
branch_author__jonathanj jonathanj
status__closed closed
resolution__fixed fixed
component__words words
keywords__IRCClient_DoS_security IRCClient DoS security
time__1302040302000000 1302040302000000
changetime__1304068909000000 1304068909000000
version__None None
owner__jonathanj jonathanj
cc__ralphm
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_5"></a>|@jonathanj commented|
|-|-|

(In [31600]) Branching to 'ctcp-dos-5029'

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_6"></a>|@jonathanj commented|
|-|-|

I've changed the CTCP dispatch code to now only reply to the first CTCP query of each type, the CTCP specification (if you can call it that) doesn't say anything about multiple queries of the same type on a single line and I don't know of any IRC clients that do this either.

I've also elected to remove replying to unknown CTCP messages which might be a debatable topic but if a proper CTCP message cannot be constructed by the sender it doesn't seem unreasonable to ignore these, considering that there is very little else we can do without negatively impacting the recipient.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @jonathanj|
|-|-|

Thanks for picking this up jonathanj.

I think the decision to only reply to one CTCP message of each type is a sensible one.  I like the implementation, too.  I think the decision not to reply to unrecognized CTCP messages at all makes sense as well.  However, I don't think we need to get rid of the ability to define a fallback handler for CTCP messages at the same time.  How about continuing to send unknown CTCP messages to `ctcpUnknownQuery`, but not doing anything in `IRCClient`'s implementation of that method?  (Deleting the method entirely would also raise a compatibility issue - it is a public API, application code might be calling it, which the change currently in the branch would break, so deleting it would require a deprecation period first).

Thanks for the other minor cleanups as well.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_8"></a>|@jonathanj commented|
|-|-|

(In [31691]) Replace the CTCP fallback handler with one that does not reply, refs [#5029](https://github.com/twisted/twisted/issues/5029).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_9"></a>|@jonathanj removed owner|
|-|-|

Replying to [exarkun](#note_7):
> However, I don't think we need to get rid of the ability to define a fallback handler for CTCP messages at the same time.  How about continuing to send unknown CTCP messages to `ctcpUnknownQuery`, but not doing anything in `IRCClient`'s implementation of that method?

Sounds good. I've left the log message in `ctcpUnknownQuery` since I thought that would be useful in the case where you're not choosing to override it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @jonathanj|
|-|-|

  1. the `test_noDuplicateCTCPDispatch` docstring suggests that no reply is made to duplicate messages, but there's no assertion about replies here.
  1. the `test_noDefaultDispatch` docstring says unrecognized messages are ignored.  It also says the fallback handler is invoked for them.  I think only one of these can be true. ;)
  1. I think duplicate unknown messages will be dispatched repeatedly because of the way the if/elif is done in `ctcpQuery`.  There's no test case to prove me wrong, at least.

Looks good otherwise!  Please merge when you've addressed the above to your satisfaction.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_11"></a>|@jonathanj commented|
|-|-|

(In [31734]) Address review commentary, refs [#5029](https://github.com/twisted/twisted/issues/5029).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_12"></a>|@jonathanj set status to `closed`|
|-|-|

(In [31735]) Merge ctcp-dos-5029.

Author: jonathanj
Reviewer: exarkun
Fixes: [#5029](https://github.com/twisted/twisted/issues/5029)

IRCClient contained a potential denial-of-service attack point where it would reply to all CTCP queries, known and unknown contained on a single line, in the form of a new line for each reply. Now only the first CTCP query of a type, per line, is replied to and unknown CTCP queries are not replied to at all.
